plutus-ledger/src/Ledger/Constraints/OffChain.hs

```diff
@@ -163,18 +174,78 @@ instance Pretty UnbalancedTx where
         , hang 2 $ vsep $ "Requires signatures:" : (pretty <$> Set.toList unBalancedTxRequiredSignatories)
         ]
 
+{- Note [Balance of value spent]
+
+The @MustSpendValue v@ constraint means: I want @v@ to be *present* in the
+transaction. That is, if @t@ is the total value produced or spent by the
+transaction, then @v Value.leq t@.
```

Comment:
Honestly, is this just too awkward a concept? It seems like it would be simpler if we had separate constraints for each side of the balance, i.e. `MustSpend` and `MustProduce`. Then it's also clearer how to "fix" the constraint if it doesn't hold.

Reply:
Yeah I suppose that could work too.

Reply:
Although it's not clear to me what we could do to fix a missing `MustSpend` constraint. We don't know which inputs to use to fix it. So we add an output with the missing value and give the tx to the wallet to fix it. Which is exactly what we do now.

Reply:
Yes, we might actually fix them both in the same way, since we can only really "make" outputs, but we can then force the wallet to balance them.

Maybe it's more that checking them is easier: you either look at the "input" side of everything, or the "output" side of everything, rather than trying to make a mixture of the two.

Reply:
Also, if we get a `MustSpend (A, 5)` constraint, and a `MustProduce (A, 5)` constraint, and there are 0 `A` tokens involved in the transaction, we still need to make sure that we only add a single output with `(A, 5)`. So we're going to continue to need the `convexUnion` function.